mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala

```diff
@@ -0,0 +1,119 @@
+ */
+@Experimental
+class AssociationRules private (
+    private var minConfidence: Double) extends Logging with Serializable {
+
+  /**
+   * Constructs a default instance with default parameters {minConfidence = 0.8}.
+   */
+  def this() = this(0.8)
+
+  /**
+   * Sets the minimal confidence (default: `0.8`).
+   */
+  def setMinConfidence(minConfidence: Double): this.type = {
+    this.minConfidence = minConfidence
+    this
+  }
+
+  /**
+   * Computes the association rules with confidence above [[minConfidence]].
+   * @param freqItemsets frequent itemset model obtained from [[FPGrowth]]
+   * @return a [[Set[Rule[Item]]] containing the assocation rules.
+   */
+  def run[Item: ClassTag](freqItemsets: RDD[FreqItemset[Item]]): RDD[Rule[Item]] = {
+    freqItemsets.flatMap { itemset =>
+      val items = itemset.items
+      items.flatMap { item =>
+        items.partition(_ == item) match {
+          // Itemsets and items in itemsets are unique, so every (antecedent, consequent) is unique
+          case (consequent, antecedent) if !antecedent.isEmpty =>
+            Some((antecedent.toSeq, (consequent.toSeq, itemset.freq)))
+          case _ => None
+        }
+      } :+ (items.toSeq, (Nil, itemset.freq))
+    }.aggregateByKey(Map[Seq[Item], Long]().empty)(
+      // Since every (antecedent, consequent) is unique, there are no collisions in the Map
+      seqOp = { case (acc, (consequent, freq)) => acc + (consequent -> freq) },
+      combOp = _ ++ _
+    ).flatMap { case (antecedent, consequentToFreq) =>
+      consequentToFreq.flatMap { case (consequent, freqUnion) =>
+        val freqAntecedent = consequentToFreq(Nil)
+        val confidence = 1.0 * freqUnion / freqAntecedent
+        if (!consequent.isEmpty && confidence >= minConfidence) {
+          Some(new Rule[Item](antecedent.toArray, consequent.toArray, freqUnion, freqAntecedent))
+        } else {
+          None
+        }
+      }
+    }
+  }
+
+  def run[Item](freqItemsets: JavaRDD[FreqItemset[Item]]): JavaRDD[Rule[Item]] = {
+    val tag = fakeClassTag[Item]
+    run(freqItemsets.rdd)(tag)
+  }
+}
+
+object AssociationRules {
+
+  /**
+   * :: Experimental ::
+   *
+   * An association rule between sets of items.
+   * @param antecedent hypotheses of the rule
+   * @param consequent conclusion of the rule
+   * @param freqUnion the frequency (num. occurrences) of the union of the antecedent and consequent
+   *                  as subsets of transactions in the transaction data
+   * @param freqAntecedent the frequency of the antecedent in the transaction data
+   * @tparam Item item type
+   */
+  @Experimental
+  class Rule[Item] private[mllib] (
+      antecedent: Array[Item],
+      consequent: Array[Item],
+      freqUnion: Double,
+      freqAntecedent: Double) extends Serializable {
+
+    def confidence: Double = 1.0 * freqUnion / freqAntecedent
+
+    require(antecedent.toSet.intersect(consequent.toSet).isEmpty, {
```

Comment:
minor: This could be quite expensive to check. It generates many temporary sets when we have many rules. Given the fact that `Rule` can be only created by `AssociationRules`, we could move this to unit test if it hurts the performance. But let's keep it here for now.